agent/consul/config_endpoint.go

```diff
@@ -389,55 +397,121 @@ func (c *ConfigEntry) ResolveServiceConfig(args *structs.ServiceConfigRequest, r
 				}
 			}
 
-			// Extract the global protocol from proxyConf for upstream configs.
-			var proxyConfGlobalProtocol interface{}
-			if proxyConf != nil && proxyConf.Config != nil {
-				proxyConfGlobalProtocol = proxyConf.Config["protocol"]
-			}
+			// First collect all upstreams into a set of seen upstreams.
+			// Upstreams can come from:
+			// - Explicitly from proxy registrations, and therefore as an argument to this RPC endpoint
+			// - Implicitly from centralized upstream config in service-defaults
+			seenUpstreams := map[structs.ServiceID]struct{}{}
 
-			// map the legacy request structure using only service names
-			// to the new ServiceID type.
 			upstreamIDs := args.UpstreamIDs
 			legacyUpstreams := false
 
+			// Before Consul namespaces were released, the Upstreams provided to the endpoint did not contain the namespace.
+			// Because of this we attach the enterprise meta of the request, which will just be the default namespace.
 			if len(upstreamIDs) == 0 {
 				legacyUpstreams = true
 
 				upstreamIDs = make([]structs.ServiceID, 0)
 				for _, upstream := range args.Upstreams {
-					upstreamIDs = append(upstreamIDs, structs.NewServiceID(upstream, &args.EnterpriseMeta))
+					sid := structs.NewServiceID(upstream, &args.EnterpriseMeta)
+					upstreamIDs = append(upstreamIDs, sid)
 				}
 			}
 
+			// First store all upstreams that were provided in the request
+			for _, sid := range upstreamIDs {
+				if _, ok := seenUpstreams[sid]; !ok {
+					seenUpstreams[sid] = struct{}{}
+				}
+			}
+
+			// Then store upstreams inferred from service-defaults
+			if serviceConf != nil && serviceConf.Connect != nil {
+				for sid := range serviceConf.Connect.UpstreamConfigs {
+					seenUpstreams[structs.ServiceIDFromString(sid)] = struct{}{}
+				}
+			}
+
+			var (
+				upstreamDefaults *structs.UpstreamConfig
+				upstreamConfigs  map[string]*structs.UpstreamConfig
+			)
+			if serviceConf != nil && serviceConf.Connect != nil {
+				if serviceConf.Connect.UpstreamDefaults != nil {
+					upstreamDefaults = serviceConf.Connect.UpstreamDefaults
+				}
+				if serviceConf.Connect.UpstreamConfigs != nil {
+					upstreamConfigs = serviceConf.Connect.UpstreamConfigs
+				}
+			}
+
+			// The goal is to flatten the mesh gateway mode in this order:
+			// 	0. Value from centralized upstream_defaults
+			// 	1. Value from local proxy registration
+			// 	2. Value from centralized upstream_configs
+			// 	3. Value from local upstream definition. This last step is done in the client's service manager.
+			var registrationMGConfig structs.MeshGatewayConfig
+
+			if args.ID != "" && args.NodeName != "" {
+				index, registration, err := state.NodeServiceWatch(ws, args.NodeName, args.ID, &args.EnterpriseMeta)
```

Comment:
This is not going to do what you expect. If the ultimate goal is to figure out what mesh gateway mode the user originally used in their explicit agent-service-register attempt, on the FIRST register this NodeServiceWatch is going to return `nil` because the catalog hasn't even seen the data yet:

order of operation:

1. user invokes agent service register
2. agent intercepts the agent/local state write (and mem write) on the client agent and uses service manager to pull resolved config data from the server via that RPC
3. when that RPC returns the agent smooshes the data from (1) with the data from (2) to make the REAL local service registration data
4. the intercept from (2) returns as if the smooshed data is what someone wrote into their operation in (1)
5. the agent writes this to agent/local state
6. time passes
7. agent does AE sync of the smooshed data only

The Servers never see what the user originally stated in (1). ServiceManager is a crazy man-in-the-middle attack on registration if that helps think about it.

The easier thing would be to pass the local mesh-gateway mode for the proxy up on the config merging RPC explicitly rather than trying to extract it from the state machine (where it doesn't exist yet).